x86: Fixes for vector masking helpers

BradleyWood opened the pull request.

BradleyWood commented:
jenkins build all

0xdaryl reviewed (changes requested):
While the changes seem reasonable on their own, I don't know what problems they are addressing.  Can you add more detail to your commit messages please?  Thanks.

BradleyWood commented:
I have updated the commit messages. @hzongaro Would you mind taking a look?

hzongaro reviewed (commented):
I think the changes look good.  There are a couple of typos in the commit message for the "x86: Fixes for vector masking helpers" commit:

"left-land side" -> "left-hand side"
"a. Perfrom" -> "a. Perform"
"I have rewritten this logic to fix logical issues simplify into 3 cases" -> "I have simpified the logic into three cases", perhaps?

This is outside the scope of this change, but may I ask you to consider adding comments to `OMR::X86::TreeEvaluator::arrayToVectorMaskHelper` in a follow up pull request?  It will certainly help anyone who is not regularly immersed in vector operations to understand what it's doing.

Finally, is it possible to add TRIL tests for `b2m`, `s2m`, `i2m` and `l2m`?  I have no idea what's supported there in terms of vector operations.

BradleyWood commented:
Updated commit messages.

> may I ask you to consider adding comments

May as well do it now. I added doxygen-style comment and add comments to explain the instruction sequence for each of relevant execution paths (AVX-512 native masking vs non-native masking).

> Finally, is it possible to add TRIL tests for b2m, s2m, i2m and l2m? I have no idea what's supported there in terms of vector operations.

I have tested these opcodes through openjdk tests. I am working on TRIL infrastructure for mask operations on another branch. Those tests require support of the complementing opcodes, such as m2v, m2b, m2s, m2i, m2l.

BradleyWood commented:
jenkins build all

hzongaro commented on compiler/x/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -6997,6 +6991,9 @@ OMR::X86::TreeEvaluator::arrayToVectorMaskHelper(TR::Node *node, TR::CodeGenerat
       TR_ASSERT_FATAL(xorEncoding != OMR::X86::Bad, "No suitable encoding form for pxor opcode");
       TR_ASSERT_FATAL(subEncoding != OMR::X86::Bad, "No suitable encoding form for psub opcode");
 
+      // Create all zero or all mask in a vector register by computing 0 - mask.
```
Typo:  "all zero or all" -> "all zero or all one" or perhaps "all zeros or all ones"

hzongaro reviewed (approved):
Looks good.  Thanks!

hzongaro commented:
x86 testing has passed.  No need to wait for the linux_riscv64 testing to complete.  Merging.
